### DISCUSSION TITLE
Loop timer + daily working periods

### DISCUSSION THREAD

**Discussion by pgorod** [arendst/Tasmota#11813] [Q&A] [answered]
Hey everyone! This is my first post here and I'd like to start by thanking everyone in this community for the wonderful Tasmota project.

I bought a few Sonoffs (Basic, TH, POW, SV) already loaded with Tasmota and I am trying to make them work for controlling my hot water heater and some other pumps. Please note that I'm not into flashing new code or re-compiling, so I'd appreciate answers that don't require that sort of intervention, _if possible_.

I have a hot-water circulation pump which I would like to work as a loop timer: on for 5 minutes, off for 30 minutes, and keep repeating this. However, there are certain times of day where I would like to stay off all the time: at night, and at certain periods where electricity is more expensive (because the pump causes other things to start working to re-heat water).

Can somebody please guide as to how I should approach this with Tasmota? I saw a post explaining how to do loops with Rules, and I see the regular timers which allow for setting certain time periods during the day. But I don't see how I could integrate the two logics...

Thanks in advance! 

Using Tasmota 9.3.1.2 on a Sonoff POW R2

**Comment by lalo-uy** [comment]
You can declare a 2nd relay on an unused gpio, an tour it on with the timer
commandas in the time of day you whant your system acrive.
Then a rule timer of 30min can tour power1 on with a Pulsetime of 5min, if
and only if Power2 is on.

**Reply by pgorod** [reply]
🤔 Let me try to understand better. Does this GPIO thing involve any wiring and soldering, or is it just something in software  I will use to configure? Like a clever way to "set a variable" in the Tasmota configuration?

And how can I know which GPIO's are unused?

Thanks a lot for your help. Sorry for being such a newbie.

**Comment by barbudor** [comment, marked as answer]
I don't see the need for a virtual relay as a Timer can trigger a rule directly.

So you have 16 timers that you can configure for for a given time of day and a patterm on week days. A Timer can then power ON, OFF or toggle but it can also trigger a Rule. Let's consider this option.
A Timer trigger would be : `on Clock#Timer=1 do ...` when `1 is the index 1..16 of the timer.
Let's say that you will have Timer #1 configured with the time of day to start the loop and Timer #2 configured with the time of day to stop it.
We are going a few tricks
1) Use PulseTime to configure the ON time duration : `PulseTime 400` will configure the ON time to 300 seconds = 5 min (see the syntax to understand why I use 300+100)
2) Use a repetitive rule every 30 minutes to (may be) turn power On. FOr that we will use variable #1 to hold the desired power state.
```
ON Clock#Minute|30 DO
  Power %var1%
ENDON
```
So every 30 minutes, this rulle will be triggered and set the power to %var1%. If %var1% is 1, it will power ON for 5 minutes (turning OFF is taken care by PulseTime)
Now to decide the time of day when you want the system should run, use Timers to trigger another rule. Let's say Timer #1 will give the start time and TImer #2 will be the end time.
```
ON Clock#Timer=1 DO
  var1 1
ENDON
ON Clock#Timer=2 DO
  var1 0
ENDON
```

All this has to be entered in the console like:
```
PulseTime 400
Rule1 ON Clock#Minutes|30 DO Power %var1% ENDON ON Clock#Timer=1 DO var1 1 ENDON ON Clock#Timer=2 DO var1 0 ENDON
Rule1 1
```

That's all
Take care to configure Timer #1 slighly before the 30 minute boundary. Like if you want the 1st Power ON to occur at 13:00, configure Timer1 to execute the rule at 12:55, so that at 13:00 when the `ON Time#Minute|30` trigger occurs, var1 is already at 1

This have 1 drawback is to performe a `Power 0` command even when you don't need it. Not a problem but will generate some MQTT messages.
If you want to avoid this, we can rework something slighly differently using RuleTimers

Hope this helps

**Reply by pgorod** [reply]
(I don't use MQTT, so I'm not worried about excess messages). Thanks for this, it looks like a great solution. I didn't know about the timers being able to trigger Rules, I guess that's the key to my problem. I will try your solution as soon as I get a chance, and I promise to come back here and leave some feedback. Thanks!

**Reply by pgorod** [reply]
I had some trouble getting this to work, I ended up tracking this to a minor bug, it should be `Time#Minute` instead of `Clock#Minute`, so the rule should look like this:

```
Rule2 ON Time#Minute|30 DO Power %var1% ENDON ON Clock#Timer=1 DO var1 1 ENDON ON Clock#Timer=2 DO var1 0 ENDON
```

I'm still playing with the Timers that trigger the rule to see how they work best.

I have a couple more questions, if you don't mind...

1. I can't do `power %var1%` directly in the Console, it doesn't evaluate the variable like it does inside the Rule. Is there any way to this in the console? A different syntax perhaps?

2. How will this scheme behave in case of power failures? Will it recover well?

Thanks!

**Reply by barbudor** [reply]
Correct, Time#Minute is the correct one. Clock#Timer for the scheduled timers

1) Indeed variables are not evaluated on console. So you can just "read" the value of var1 by typing `var1` + return in the console. It should be 0 or 1
2) By default at restart var1 is empty so the command `Power %var1%` will do nothing until the next Timer event that will set it to 1.
An easy way of sorting that would be to use `%mem1%` instead of `%var1%` as Mem are written to flash. However that means that flash will be written twice a day which may reduce its lifetime.

**Reply by sfromis** [reply]
If you'd like to have a console command including the use of %var1%, you could put it into something like `on event#mycommand do var9 %var1% %value% endon` and invoke it using `event mycommand=123`. While flash wear from mem writes can be a real issue, it should not be too bad with twice a day - unless you plan for a human lifetime of operation.

**Reply by pgorod** [reply]
Thanks to both of you. I think I can avoid the use of flash by adding this rule:

`ON system#boot DO var1 1 ENDON`

Power outages aren't frequent, and I prefer the `1` state as a default when power comes back up. It may waste a bit of energy circulating water during the middle of the night, but at least I won't have a bunch of elderly people waking up to a cold shower... 😅
